providers/agent/mcorpc/audit/choria.go

```diff
@@ -67,13 +78,30 @@ func Request(request protocol.Request, agent string, action string, data json.Ra
 	mu.Lock()
 	defer mu.Unlock()
 
-	f, err := os.OpenFile(logfile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0600)
+	f, err := os.OpenFile(logfile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, fs.FileMode(logfileMode))
 	if err != nil {
 		log.Warnf("Auditing is not functional because opening the logfile '%s' failed: %s", logfile, err)
 		return false
 	}
 	defer f.Close()
 
+	if logfileGroup != "" {
+		grp, err := user.LookupGroup(logfileGroup)
+		if err == nil {
+			gid, err := strconv.Atoi(grp.Gid)
+			if err == nil {
+				err = os.Chown(logfile, os.Getuid(), gid)
+				if err != nil {
+					log.Warnf("Unable to chown file %s to group %s: %s", logfile, logfileGroup, err)
+				}
+			} else {
+				log.Warnf("Unable to convert group %s GID: %s", logfileGroup, err)
+			}
```

Comment:
You cn void these nested ifs by making a function that creates the file, then you can just return early and so avoid all the else stuff.

This is actually quite difficult to do right, since in this code your file will exist for a short while as the wrong user - open as root then chown - this means your file tailers could fail for a while and depending on how tjhey handle that scenario could prevent the file from being accessed.

Safer to make a tmp file in the same directory (has to be the same directory) chown the temp file then rename it and pass back the file handle

This means opening the file becomes:

 1. Open for append and set the file handle
 2. If that fails, call into the create function that returns the new file handle
 3. Log to the file handle

This avoids the race and the file only ever exist in the intended name with the right permissions and ownership

Reply:
Both Splunk Forwarder and Vector can handle brief periods of file inaccessibility as they will retry if the file is created and momentarily the wrong GID.  This is almost certainly a non issue for most file tailers since none can guarantee a file ever exists so they almost always will retry to access a file since things like OS boot it's not guaranteed a file will exist yet so the tail programs are almost always designed to retry.

Is it worth the extra logic to handle something that could very well be a non-issue?  Since the audit log won't even exist until a request comes through, any tailer being used with Choria would have to be able to handle the absence of the file and retry and absence and lack of read permission will generally look the same to programs going to open the file.

Reply:
No you're right thats ok, lets try to remove the nested ifs though